### ISSUE TITLE
Failed to create a temp. directory for the wkhtmltopdf binary.

### ISSUE THREAD

**Issue by daminimehra28** [todd-elvers/html-to-pdf-converter#1]
I used this library to convert HTML string to pdf. but it gives this error
java.lang.RuntimeException: Failed to create a temp. directory for the wkhtmltopdf binary.
        at te.htmltopdf.HtmlToPdfBinaryExtractor.lambda$createNewTempFile$1(HtmlToPdfBinaryExtractor.java:53)
        at te.htmltopdf.-$$Lambda$HtmlToPdfBinaryExtractor$Vns0Vitn0CPI5a6c7CA39QRDFg8.apply(Unknown Source:2)
        at io.vavr.control.Try$-CC.$default$getOrElseThrow(Try.java:684)
        at io.vavr.control.Try$Failure.getOrElseThrow(Unknown Source:0)
        at te.htmltopdf.HtmlToPdfBinaryExtractor.createNewTempFile(HtmlToPdfBinaryExtractor.java:53)
        at te.htmltopdf.HtmlToPdfBinaryExtractor.writeBinaryToTempDir(HtmlToPdfBinaryExtractor.java:40)
        at te.htmltopdf.HtmlToPdfBinaryExtractor.extract(HtmlToPdfBinaryExtractor.java:25)
        at te.htmltopdf.HtmlToPdfBinaryResolver.resolveBinaryFile(HtmlToPdfBinaryResolver.java:45)
        at te.htmltopdf.HtmlToPdfBinaryResolver.resolve(HtmlToPdfBinaryResolver.java:31)
        at te.htmltopdf.HtmlToPdfFileConverter.<init>(HtmlToPdfFileConverter.java:34)

**Comment by todd-elvers** [comment]
So as the error message states the library is trying to create a temporary directory to store the `wkhtmltopdf` binary, which is used to convert HTML to PDF.  I will attempt to track down this error today, but as a workaround to disable the binary extraction logic you can:

- Download the `wkhtmltopdf` binary from their website (or extract it from this JAR)
- Set the environment variable `WKHTMLTOPDF_BINARY` to the absolute path to your `wkhtmltopdf` binary

**Comment by daminimehra28** [comment]
Hello,
Thanks for getting back to me.


I have one more question how to set environment varibale? .let me know as
soon as possible.

**Comment by todd-elvers** [comment]
What operating system are you using?
What version of Java are you using?

How to set your environment variables depends on the operating system you are using. Are you familiar with how to set your PATH variable or your JAVA_HOME variable?  It's very similar to that.  Google your operating system and "how to set environment variables" and you'll find a good answer.

As an example of what it should be:  
If you're on Windows and have the `wkhtmltopdf.exe` binary in `C:\Users\Todd\Desktop` then setting `WKHTMLTOPDF_BINARY` to `C:\Users\Todd\Desktop\wkhtmltopdf.exe` should cause your binary to be used instead of the one bundled in the library.

**Comment by daminimehra28** [comment]
thanks for your reply.
OK let me try this and let you know. I tried your solution but my android application crashed. I set the environment variable.

**Comment by todd-elvers** [comment]
Okay, I'll keep an eye on this post for your response.

I'm also updating the library to be a little simpler in terms of temporary file writing.
Previous logic: create a directory in the temp directory and copy the binary into there 
New logic: Copy the binary directly to the temp directory (no directory creation anymore).

This simplifies the binary extraction logic and _might_ help with your issue.  I'll be releasing the next version today and when I do I'll let you know.

If you get some time please let me know how you're using this library and with what operating system & tech stack so I can better test your use cases.  For example you mentioned Android and that environment might not let you create temporary files in the same way.

**Comment by daminimehra28** [comment]
Hi thanks for ur response, I had added it as a library in my Android
project. For example in app level gradle file I add
"implementation xyz library "

And run my Android application..
If I need to add any other configuration let me know.
Tomorrow I will send you my Android project code snippet.

**Comment by todd-elvers** [comment]
@daminimehra28 I have received your code snippets and I'm reviewing them now.  I'll also look into whether or not Android wants you to create temporary files another way.  I'll get back to you tomorrow and I'll hopefully have a release for you by Monday.

**Comment by todd-elvers** [comment]
@daminimehra28 Below is a breakdown of what I understood from you, the code, and some googling I did.

### Your Situation
You're trying to use this library to generate PDFs on Android OS.  You're providing the String HTML and the output file you'd like the PDF to be written to.  However when you attempt to perform the conversion you get an exception related to a binary failing to be written to a temporary file.

### How this library works
This library is simply a wrapper around the `wkhtmltopdf` program; it takes your HTML along with an output file and passes them to the `wkhtmltopdf` program which itself creates the PDF and writes it to your output file.  To accomplish this the correct `wkhtmltopdf` binary must be available, so I bundled all the different versions of it into this library and just cached it to a temporary file on the first call of this library.

### The Problem
The exact problem you're experiencing is likely due to a design flaw in how my library extracts the `wkhtmltopdf` binary, which probably causes some permissions issues in more heavily restricted environments (like Android).  I have identified a way to simplify the binary extraction process which should resolve this issue.  But...

### The Bigger Problem
`wkhtmltopdf` binary only supports Windows, MacOS, and Linux.  That means even after I fix the temporary file issue this library still won't generate PDFs for you on Android.

### Other HTML-to-PDF Solutions
If you're running Android OS API level 19 you can use their [PrintedPdfDocument](https://developer.android.com/reference/android/print/pdf/PrintedPdfDocument.html) class to generate PDFs from HTML.  Google some examples of that class and I'm sure you'll find your use case in the results somewhere.

The above is the preferred way to generate PDFs on Android but if you're not on level 19 yet here is a good [stackoverflow question with several additional options](https://stackoverflow.com/questions/2499960/how-to-create-pdfs-in-an-android-app).

Another option would to use some cloud service and just HTTP POST your String to it.  [Docmosis](https://www.docmosis.com/) is one cloud service that performs PDF generation.


**In Conclusion**
I'm sorry this library couldn't help you.  You did help me find a design flaw in my library so thank you for taking the time to use this and to create an issue when you found a problem with it.  I'll leave this issue open for a while so you can respond or ask any further questions you may have.
